File: bindings/rsac-napi/tests/capture_smoke.mjs
// Deterministic-route SystemDefault capture smoke for @rsac/audio (rsac-3635).
// Linux-CI-only. Invoked EXPLICITLY (node --test tests/capture_smoke.mjs) — the
// filename has NO `.test.` infix, so `npm test` (`node --test`, which auto-globs
// *.test.mjs) in the device-free check-bindings job does NOT pick it up.
// Defense-in-depth: still self-skips when RSAC_CI_AUDIO_DETERMINISTIC is unset.
//
// Mirrors the Rust system-capture contract (tests/ci_audio/system_capture.rs +
// lifecycle_terminal.rs): build -> start -> non-silent chunk -> stop ->
// terminal-observable read -> (implicit clean drop). RMS floor 0.01 copied from
// helpers::verify_rms_energy. Per index.d.ts: the non-blocking read() is NOT
// terminal-observable (throws a recoverable "not running" while stopped), so we
// poll it for the non-silent chunk; readBlocking() IS terminal-observable, so it
// is the surface we assert throws after stop().
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const rsac = require('../index.js');

const DET = process.env.RSAC_CI_AUDIO_DETERMINISTIC === '1';
const timeoutMs = Number(process.env.RSAC_TEST_CAPTURE_TIMEOUT_SECS ?? '15') * 1000;

test('SystemDefault capture delivers non-silence and stops cleanly', { skip: !DET && 'RSAC_CI_AUDIO_DETERMINISTIC != 1' }, async () => {
  const cap = rsac.AudioCapture.create(rsac.CaptureTarget.systemDefault(), 48000, 2);
  cap.start();
  assert.equal(cap.isRunning, true, 'isRunning after start()');

  let gotNonSilence = false;
  const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    let chunk = null;
    try {
      chunk = cap.read();
    } catch {
      // read() is not terminal-observable: while running it may throw a
      // transient recoverable "not running" hiccup — tolerate and retry.
    }
    if (chunk && chunk.rms > 0.01) {
      gotNonSilence = true;
      break;
    }
    // 10ms backoff between empty polls — matches the Python (sleep 0.01)
    // and Rust (10ms) smoke loops; a tight loop pegs a runner core.
    await sleep(10);
  }
  assert.ok(gotNonSilence, 'deterministic source: at least one non-silent chunk');

  cap.stop();
  assert.equal(cap.isRunning, false, 'isRunning false after stop()');
  // readBlocking() is terminal-observable — throws the terminal error after stop.
  assert.throws(() => cap.readBlocking(), 'readBlocking() after stop() must throw terminal');
});

File: bindings/rsac-python/tests/capture_smoke.py
"""Deterministic-route SystemDefault capture smoke for the rsac Python binding
(rsac-3635). Linux-CI-only: run under the deterministic PipeWire route with
RSAC_CI_AUDIO_DETERMINISTIC=1. Skips cleanly (exit 0) otherwise.

Invoked BY EXPLICIT PATH only (like the device-free smoke.py), so it never runs
in the headless check-bindings job:

    python bindings/rsac-python/tests/capture_smoke.py

Mirrors smoke.py's failures[] + sys.exit(1) discipline, and mirrors the Rust
system-capture contract in tests/ci_audio/system_capture.rs +
lifecycle_terminal.rs (build -> start -> non-silent buffer -> stop ->
terminal-observable read -> clean drop), with the RMS floor copied from
helpers::verify_rms_energy (0.01)."""

import os
import sys
import time

import rsac

DET = os.environ.get("RSAC_CI_AUDIO_DETERMINISTIC") == "1"
if not DET:
    print("capture smoke SKIPPED: RSAC_CI_AUDIO_DETERMINISTIC != 1")
    sys.exit(0)

timeout = float(os.environ.get("RSAC_TEST_CAPTURE_TIMEOUT_SECS", "15"))
failures = []

# 1. build a SystemDefault capture (48000 Hz, 2ch). Under the deterministic
#    route a build/start failure is a hard fail.
cap = rsac.AudioCapture(rsac.CaptureTarget.system_default(), sample_rate=48000, channels=2)

# 2. start; is_running must be True.
cap.start()
assert cap.is_running, "capture must be running after start()"

# 3. read loop until a non-silent buffer OR the bounded deadline. Under the
#    deterministic route at least one non-silent buffer is mandatory.
got_non_silence = False
deadline = time.monotonic() + timeout
while time.monotonic() < deadline:
    buf = cap.try_read()
    if buf is None:
        time.sleep(0.01)
        continue
    if buf.rms() > 0.01:  # mirrors helpers::verify_rms_energy floor
        got_non_silence = True
        print(f"first non-silent buffer: {buf.num_frames} frames RMS={buf.rms():.6f}")
        break

if not got_non_silence:
    failures.append("deterministic source: no non-silent buffer within timeout")

# 4. stop; is_running must be False.
cap.stop()
if cap.is_running:
    failures.append("is_running must be False after stop()")

# 5. terminal-observable: blocking read() after stop() raises the stream's true
#    FATAL terminal (StreamEnded → StreamError with "Stream ended: ..."), not a
#    recoverable "not running" downgrade (the rsac-477d regression class).
try:
    cap.read()
    failures.append("read() after stop() should raise the terminal error")
except rsac.StreamError as e:
    if "stream ended" in str(e).lower():
        print("post-stop read raised the fatal StreamEnded terminal: ok")
    else:
        failures.append(
            f"read() after stop() raised StreamError but not the StreamEnded "
            f"terminal (got: {e}) — recoverable downgrade regression?"
        )
except rsac.RsacError as e:
    failures.append(
        f"read() after stop() raised a non-Stream RsacError: {type(e).__name__}: {e}"
    )

# 6. close/drop cleanly — close() is idempotent.
cap.close()
cap.close()

if failures:
    for f in failures:
        print(f"FAIL: {f}", file=sys.stderr)
    sys.exit(1)
print("python capture smoke: OK")
